src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -446,54 +462,103 @@ private Task WriteFormattedInt32Async(int value, CancellationToken cancellationT
+                    throw CreateOperationCanceledException(_pendingException ?? error, cancellationToken);
+                }
+                else if (_pendingException != null)
                 {
                     // If we incurred an exception in non-linear control flow such that
                     // the exception didn't bubble up here (e.g. concurrent sending of
                     // the request content), use that error instead.
                     throw new HttpRequestException(SR.net_http_client_execution_error, _pendingException);
                 }
-
-                // Otherwise, propagate this exception, wrapping it if necessary to
-                // match exception type expectations.
-                if (error is InvalidOperationException || error is IOException)
+                else if (error is InvalidOperationException || error is IOException)
                 {
+                    // If it's an InvalidOperationException or an IOException, for consistency
+                    // with other handlers we wrap the exception in an HttpRequestException.
                     throw new HttpRequestException(SR.net_http_client_execution_error, error);
                 }
-                throw;
+                else
+                {
+                    // Otherwise, just allow the original exception to propagate.
+                    throw;
+                }
             }
         }
 
+        private CancellationTokenRegistration RegisterCancellation(CancellationToken cancellationToken)
+        {
+            // Cancellation design:
+            // - We register with the SendAsync CancellationToken for the duration of the SendAsync operation.
+            // - We register with the Read/Write/CopyToAsync methods on the response stream for each such individual operation.
+            // - The registration disposes of the connection, tearing it down and causing any pending operations to wake up.
+            // - Because such a tear down can result in a variety of different exception types, we check for a cancellation
+            //   request and prioritize that over other exceptions, wrapping the actual exception as an inner of an OCE.
+            // - A weak reference to this HttpConnection is stored in the cancellation token, to prevent the token from
+            //   artificially keeping this connection alive.
+            return cancellationToken.Register(s =>
+            {
+                var weakThisRef = (WeakReference<HttpConnection>)s;
+                if (weakThisRef.TryGetTarget(out HttpConnection strongThisRef))
+                {
+                    if (NetEventSource.IsEnabled) strongThisRef.Trace("Cancellation requested. Disposing of the connection.");
+                    strongThisRef.Dispose();
+                }
+            }, _weakThisRef);
+        }
+
+        private static bool ShouldWrapInOperationCanceledException(Exception error, CancellationToken cancellationToken) =>
```

Comment:
This seems like a good reason to change Sockets so that they consistently throw OperationCancelledException on Close.  That seems like it would save us a fair amount of code here, right?

Reply:
I'm not sure how we do it consistently.  If you Dispose of the socket and then call a method on it, getting an OperationCanceledException rather than an ObjectDisposedException seems... odd.  It'll also require more try/catch/throws as we'd need to translate ObjectDisposedExceptions from SafeHandles into OperationCanceledExceptions.

Reply:
Yeah, that makes sense.  So we will always have to translate ObjectDisposedException to OperationCancelledException.

What we could (and probably should) do, I think, is change Socket behavior to be consistent about throwing OperationCancelled instead of returning 0 bytes.  This would simplify things a bit here, right?

Reply:
> What we could (and probably should) do, I think, is change Socket behavior to be consistent about throwing OperationCancelled instead of returning 0 bytes. This would simplify things a bit here, right?

Does the socket know the difference to be able to make that call?

Reply:
I think it should be able to.  We can do what we're doing here: track that we're closing, and if the operation returns EOF, we just treat it as a cancelled op.

Changing Sockets to be consistent would be good regardless, if we can do it.

Reply:
But we're not passing the cancellation token to the socket, so it doesn't know if cancellation was requested.

Reply:
The latter, yeah.

Maybe I'm remembering wrong, but I seem to recall that the issue here is that Linux behaves different than Windows: Windows throws an error, Linux just returns EOF.  So the goal here would be to get consistent, and also to simplify HttpClient usage here.

I could be wrong though... you've been testing this scenario, what have you found?

Reply:
> you've been testing this scenario, what have you found?

I believe that lines up with what I saw... I had to add the EOF handling once I started testing on Linux (though I can't say for sure I never saw it on Windows).

Reply:
https://github.com/dotnet/corefx/issues/27066